src/main/kotlin/pet/project/app/util/ProfilingAnnotationBeanPostProcessor.kt

```diff
@@ -0,0 +1,88 @@
+package pet.project.app.util
+
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.config.BeanPostProcessor
+import org.springframework.cglib.proxy.Enhancer
+import org.springframework.cglib.proxy.MethodInterceptor
+import org.springframework.cglib.proxy.MethodProxy
+import org.springframework.context.annotation.Configuration
+import pet.project.app.annotation.Profiling
+import java.lang.reflect.Method
+import kotlin.reflect.KClass
+import kotlin.reflect.full.memberProperties
+import kotlin.reflect.full.primaryConstructor
+import kotlin.reflect.full.superclasses
+import kotlin.reflect.jvm.javaField
+import kotlin.reflect.jvm.jvmErasure
+
+
+@Configuration
+class ProfilingAnnotationBeanPostProcessor : BeanPostProcessor {
+
+    private val beanClassesForProfilingMap = hashMapOf<String, KClass<out Any>>()
+
+    override fun postProcessBeforeInitialization(bean: Any, beanName: String): Any? {
+        val beanClass: KClass<out Any> = bean::class
+        if (annotationExistsRecursively(beanClass, Profiling())) {
+            log.info("PROFILING: Remembering {} during 'before init'", beanName)
+            beanClassesForProfilingMap[beanName] = beanClass
+        }
+        return bean
+    }
+
+    private fun annotationExistsRecursively(beanClass: KClass<out Any>, annotation: Annotation): Boolean {
+        return beanClass.annotations.contains(annotation)
+                || beanClass.superclasses.any { annotationExistsRecursively(it, annotation) }
+    }
+
+    @Suppress("ReturnCount")
+    override fun postProcessAfterInitialization(bean: Any, beanName: String): Any? {
+        val beanClass = beanClassesForProfilingMap[beanName] ?: return bean
+
+        val enhancer = Enhancer()
+        enhancer.setSuperclass(beanClass.java)
+        enhancer.setCallback(MethodInterceptor(::invokeWithProfiling))
+
+        val constructorParams: Array<Class<*>> = getConstructorParams(beanClass) ?: return enhancer.create()
+        constructorParams.forEach { log.info("constructor param: {}", it) }
```

Comment:
logging in loop is bad practice. Please, log params in the next way:
```suggestion
        log.info("constructor params: {}", constructorParams.contentToString())
```

Reply:
By the way, do we really need to log the parameters?